features/steps/test_behave.py

```diff
@@ -1,108 +1,215 @@
+            assert exit_code != 0
+        elif action == 'autoremove':
+            subprocess.check_call(['dnf', '-y', action], stdout=subprocess.PIPE)
+        else:
+            execute_dnf_command([action] + split(pkg), context.repo)
 
-  if manager == 'rpm':
-    execute_rpm_command(split(pkg), action)
-  elif manager == 'dnf':
-    execute_dnf_command([action] + split(pkg), 'test')
 
 @then('package "{pkg}" should be "{state}"')
 def then_package_state(context, pkg, state):
-  assert state in ["installed", "removed", "absent"]
-  assert pkg
-  pkgs = get_package_list()
-  assert pkgs
-  removed, installed = diff_package_lists(context.pre_packages, pkgs)
-  assert removed != None and installed != None
+    assert state in ["installed", "removed", "absent", "upgraded", 'unupgraded', "downgraded", 'present', 'upgraded-to']
+    assert pkg
+    pkgs = get_package_list()
+    pkgs_ver = get_package_version_list()
+    pkgs_dnf_ver = get_package_dnf_version_list()
+    assert pkgs
+    assert context.pre_packages
+    removed, installed = diff_package_lists(context.pre_packages, pkgs)
+    assert removed is not None and installed is not None
   
-  for n in split(pkg):
-    if state == 'installed':
-      assert ('+' + n) in installed
-      installed.remove('+' + n)
-    if state == 'removed':
-      assert ('-' + n) in removed
-      removed.remove('-' + n)
-    if state == 'absent':
-      assert ('+' + n) not in installed
-      assert ('-' + n) not in removed
-
-  ''' This checks that installations/removals are always fully specified,
-  so that we always cover the requirements/expecations entirely '''
-  if state != 'absent':
-    assert not installed and not removed
+    for n in split(pkg):
+        if state == 'installed':
+            assert ('+' + n) in installed
+            installed.remove('+' + n)
+            post_present = package_version_lists(n, pkgs_ver)
+            assert post_present
+            post_dnf_present = package_version_lists(n, pkgs_dnf_ver)
+            assert post_dnf_present
+        if state == 'removed':
+            assert ('-' + n) in removed
+            removed.remove('-' + n)
+            post_absence = package_absence(n, pkgs_ver)
+            assert not post_absence
+            post_dnf_absence = package_absence(n, pkgs_dnf_ver)
+            assert not post_dnf_absence
+        if state == 'absent':
+            assert ('+' + n) not in installed
+            assert ('-' + n) not in removed
+            post_absence = package_absence(n, pkgs_ver)
+            assert not post_absence
+            post_dnf_absence = package_absence(n, pkgs_dnf_ver)
+            assert not post_dnf_absence
+        if state == 'upgraded':
+            pre_ver = package_version_lists(n, context.pre_packages_version)
+            post_ver = package_version_lists(n, pkgs_ver)
+            assert post_ver
+            assert pre_ver
+            assert post_ver > pre_ver
+        if state == 'unupgraded':
```

Comment:
That is state when package was installed before test and after test is still present with same version, release.